### PR TITLE
add improved boundary check and check malloc return is NULL

### PR DESCRIPTION
**Pull request by yskcg**
### Commit Summary

* add the malloc return and free pointer check
* add the boundary check ,when parse the HTTP protocol headers

### File Changes

* M `example/ndpiReader.c`
* M `src/lib/ndpi_main.c`

### Why changes

* valgrind memcheck give the warnings

### CODE REVIEW

File: example/ndpiReader.c [hunk 1]
```diff
@@ -3047,7 +3050,10 @@ static void ndpi_process_packet(u_char *args,
      Leave the free as last statement to avoid crashes when ndpi_detection_giveup()
      is called above by printResults()
   */
-  free(packet_checked);
+  if(packet_checked){
```
Comment (utoni) [inline example/ndpiReader.c:3053]: NUL check already done

File: src/lib/ndpi_main.c [hunk 2]
```diff
@@ -4928,7 +4928,7 @@ void ndpi_parse_packet_line_info(struct ndpi_detection_module_struct *ndpi_str,
   packet->line[packet->parsed_lines].len = 0;
 
   for (a = 0; ((a+1) < packet->payload_packet_len) && (packet->parsed_lines < NDPI_MAX_PARSE_LINES_PER_PACKET); a++) {
-    if((packet->payload[a] == 0x0d) && (packet->payload[a+1] == 0x0a)) {
+    if(((a + 1) < packet->payload_packet_len) &&(packet->payload[a] == 0x0d) && (packet->payload[a+1] == 0x0a)) {
```
Comment (utoni) [inline src/lib/ndpi_main.c:4931]: Isn't it already checked in the loop condition?

**Review by utoni — CHANGES_REQUESTED** [review]

**Comment by lucaderi** [comment]
Thanks, nice checks but don't forget that this is just a demo app

File: src/lib/ndpi_main.c (hunk 2 above)
Comment (yskcg) [reply src/lib/ndpi_main.c:4931]: Yeah，U are right,I don't attention the loop case;

**Pull request merged**